promql/prettier/prettier.go

```diff
@@ -0,0 +1,459 @@
+		isNewLineApplied:   true,
+		expectNextRHS:      false,
+		previousBaseIndent: "  ", // start with initial indent as previous val to skip new line at start.
+	}
+}
+
+func (p *Prettier) prettifyItems(items []parser.Item, index int, result string) string {
+	item := items[index]
+	value := item.Val
+	switch item.Typ {
+	case parser.LEFT_PAREN:
+		result += p.pd.apply() + value + "\n" + p.pd.inc(1).resume()
+	case parser.RIGHT_PAREN:
+		result += "\n" + p.pd.dec(1).apply() + value
+	case parser.LEFT_BRACE, parser.LEFT_BRACKET:
+		result += value + "\n" + p.pd.inc(1).resume()
+	case parser.RIGHT_BRACE, parser.RIGHT_BRACKET:
+		result += p.pd.dec(1).apply() + value
+	case parser.IDENTIFIER, parser.DURATION, parser.NUMBER:
+		result += p.pd.apply() + value
+	case parser.STRING:
+		result += value + ",\n"
+	case parser.SUM:
+		result += p.pd.apply() + value
+	case parser.COLON, parser.EQL, parser.EQL_REGEX, parser.NEQ, parser.NEQ_REGEX:
+		result += value
+	case parser.ADD, parser.SUB, parser.DIV, parser.MUL:
+		result += "\n" + p.pd.dec(1).apply() + value + "\n" + p.pd.inc(1).resume()
+	case parser.GTR, parser.BOOL:
+		if items[index+1].Typ == parser.BOOL {
+			index++
+			result += "\n" + p.pd.dec(1).apply() + value + " " + items[index].Val + "\n" + p.pd.inc(1).resume()
+		} else {
+			result += "\n" + p.pd.dec(1).apply() + value + "\n" + p.pd.inc(1).resume()
+		}
+	case parser.COMMENT:
+		result += p.pd.apply() + value + "\n"
+
+	case parser.BLANK:
+	}
+	if item.Typ == parser.EOF || len(items) == index+1 {
+		return p.removeTrailingLines(result)
+	}
+	return p.prettifyItems(items, index+1, result)
+}
+
+var exmpt = []uint{
+	parser.LEFT_BRACE, parser.EQL,
+}
+
+func (p *Prettier) prettify(items []parser.Item, index int, result string) (string, error) {
+	var (
+		baseIndent string
+		item       = items[index]
+	)
+	if item.Typ == parser.EOF {
+		return result, nil
+	}
+	if !p.exmptBaseIndent(item) {
+		baseIndent, _, _ = p.getBaseIndent(item, p.Node)
+		if baseIndent != p.pd.previousBaseIndent { // denotes change of node
+			result += "\n"
+		}
+		result += baseIndent
+		p.pd.previousBaseIndent = baseIndent
+	}
+
+	switch item.Typ {
+	case parser.LEFT_PAREN:
+		result += item.Val
+	case parser.LEFT_BRACE:
+		result += item.Val + p.pd.insertNewLine()
+		p.pd.expectNextIdentifier = "label"
+	case parser.RIGHT_BRACE:
+		result += p.pd.previousBaseIndent + item.Val
+		p.pd.expectNextIdentifier = ""
+	case parser.IDENTIFIER:
+		if p.pd.expectNextIdentifier == "label" {
+			result += p.pd.apply()
+		}
+		result += item.Val
+	case parser.STRING:
+		result += item.Val + ",\n"
+	case parser.RIGHT_PAREN:
+		result += p.pd.dec(1).apply() + item.Val
+		p.pd.expectNextIdentifier = ""
+	case parser.EQL:
+		result += item.Val
+	case parser.ADD:
+		result += item.Val
+	case parser.COMMENT:
+		result += p.pd.previousBaseIndent + item.Val + "\n"
+	}
+	if index+1 == len(items) {
+		return result, nil
+	}
+	ptmp, err := p.prettify(items, index+1, result)
```

Comment:
I think keeping a lot of variables in the starting of a function will not be good, so turned out to be recursive. The reason for tail-recursive is to make the functional length short, else it will lead to a recursive call in each case block, thus making the code look untidy.